### ISSUE TITLE
Order by Guid.NewGuid() returns same order every time

### ISSUE THREAD

**Issue by dylanparry** [aspnet/EntityFramework#2069]
I am attempting to order my table in a random order so that I can select a random row from it. The problem appears to be that the ```OrderBy(…)``` part of the Linq statement is being ignored and no SQL is generated for it.

Here’s what I’m using:

```csharp
var items = await this.context.Items.OrderBy(a => Guid.NewGuid()).FirstAsync();
```

And here’s the SQL that’s being generated for it:

```sql
exec sp_executesql N'SELECT TOP(@p0) [a].[ID], … [other columns omitted]
FROM [Items] AS [a]',N'@p0 int',@p0=1
```

As you can see there’s no ```ORDER BY``` being appended to the SQL, which results in the same "random" row being returned every single time.

I had originally intended on using paging with ```.Skip(…).FirstAsync()``` to select a random row, but as I’m stuck using SQL Server 2008 for the foreseeable future, that doesn’t work either (See https://github.com/aspnet/EntityFramework/issues/1326)

Can anyone confirm that the ```ORDER BY``` isn’t being appended for other cases, and if so is there a workaround?

Thanks.

**Comment by divega** [comment]
@maumar this issue sounds somewhat familiar. Any chance we have a fix already or was it something different?

**Comment by maumar** [comment]
Currently we will produce the following query:

```
SELECT TOP(1) [a].[Id], [a].[Name]
FROM [Items] AS [a]
ORDER BY @__NewGuid_0
```

which will fail with the following exception:
Additional information: The SELECT item identified by the ORDER BY number 1 contains a variable as part of the expression identifying a column position. Variables are only allowed when ordering by an expression referencing a column name.

This is because we funcletize Guid.NewGuid() into a constant. In order for this to work we would have to recognize the method and send it directly to the server, i.e. 

```
SELECT TOP(1) [a].[Id], [a].[Name]
FROM [Items] AS [a]
ORDER BY NEWID()
```

**Comment by maumar** [comment]
Per conversation with Diego and Andrew: plan is to disable partial evalulation for methods and delegate calls and attempt to translate some methods to SQL (e.g. Guid.NewGuid() -> NEWID() )

Users would be able to opt-in to the partial eval by passing the result of the method as a closure, like so:

```
var myGuid = Guid.NewGuid();
await this.context.Items.OrderBy(a => myGuid).FirstAsync();
```

**Comment by maumar** [comment]
Fixed in 3f0fc0baeb28eba3b3bbadb69a4a87e24fabb680
